### PR TITLE
[5.0] extend life of `http_plugin_state` to ensure no invalid memory accesses during nodeos shutdown

### PR DESCRIPTION
**Pull request by spoonincode**
During the destruction of `beast_http_session`,
https://github.com/AntelopeIO/leap/blob/04774eb7726ae95a6cb795b493fcf0f25021bc5f/plugins/http_plugin/include/eosio/http_plugin/beast_http_session.hpp#L270-L281
there are accesses to `plugin_state_` which is simply a reference to an object that lives in `http_plugin_impl`,
https://github.com/AntelopeIO/leap/blob/04774eb7726ae95a6cb795b493fcf0f25021bc5f/plugins/http_plugin/include/eosio/http_plugin/beast_http_session.hpp#L86
Additionally, notice that `beast_http_session`'s `socket_` has been constructed with the io_context from `http_plugin_impl::plugin_state`'s `thread_pool`. So destruction of the `socket_` will also require valid access to the `plugin_state` object.

But it is possible for a `beast_http_session` to outlive `http_plugin` which causes these accesses during destruction to be invalid. Consider a situation where appbase's executor (`priority_queue_executor`) has a queued callback holding the sole reference count of a `shared_ptr<beast_http_session>` and appbase is commanded to `quit()`. appbase's `exec()` upon exiting its run loop will then call appbase's `shutdown()`. `application_base::shutdown()` will then effectively call `http_plugin::plugin_shutdown()` (stopping http threads) and then destroy `http_plugin`. However, `priority_queue_executor` will still be keeping alive a `beast_http_session` via some pending callback. As appbase continues to destruct eventually `priority_queue_executor` is destroyed which destroys all pending callbacks. It's at that time the last reference to the `shared_ptr<beast_http_session>` is removed causing it to be destroyed and expecting access to `http_plugin_impl::plugin_state`.

But wait! When appbase's io_context stops it first clears its pending queue before `application_base::shutdown()`, right?
https://github.com/AntelopeIO/appbase/blob/b75b31e14f966fa3de6246e120dcba36c6ce5264/include/appbase/application_base.hpp#L148-L150
afaict this is racy: there is nothing to stop code in some other non-main thread (like http_plugin's threads!) from queueing more callbacks on to appbase's executor between the time `exec.clear()` is called and, say, `http_plugin::shutdown()` is called which halts its threads.

Not immediately seeing any good approaches to resolve this issue (especially in a patch release) beyond migrating `http_plugin_impl::plugin_state` to a `shared_ptr` so that every `beast_http_session` can keep it alive as long as it needs to. The `socket_` needing `http_plugin_impl::plugin_state`'s `thread_pool` to remain alive is seemingly what makes this non-negotiable without more significant refactoring.

I am considering this resolving #2022 because while I can't get the exact fatal malloc error to print, the failure I can reproduce here is at the same point in time.

Was unable to reproduce the issue in 3.2 or 4.0. Curiously, those versions already do have a `std::shared_ptr<http_plugin_state>` so maybe that's why.

### CODE REVIEW

File: plugins/http_plugin/include/eosio/http_plugin/beast_http_session.hpp [hunk 1]
```diff
@@ -251,13 +252,13 @@ class beast_http_session : public detail::abstract_conn,
    // shared_from_this() requires default constructor
    beast_http_session() = default;
 
-   beast_http_session(Socket&& socket, http_plugin_state& plugin_state, std::string remote_endpoint,
+   beast_http_session(Socket&& socket, std::shared_ptr<http_plugin_state> plugin_state, std::string remote_endpoint,
                       api_category_set categories, const std::string& local_address)
-       : socket_(std::move(socket)), categories_(categories), plugin_state_(plugin_state),
+       : plugin_state_(plugin_state), socket_(std::move(socket)), categories_(categories),
```
Comment (heifner) [inline plugins/http_plugin/include/eosio/http_plugin/beast_http_session.hpp:257]: This can be `std::move(plugin_state)`

**Review by heifner — APPROVED** [review]

**Review by linh2931 — APPROVED** [review]
Like your detailed analysis.

**Review by greg7mdp — APPROVED** [review]

**Comment by ericpassmore** [comment]
Note:start
group: STABILITY
category: INTERNALS
summary: Extend life of `http_plugin_state` to ensure no invalid memory accesses during nodeos shutdown.
Note:end

File: plugins/http_plugin/include/eosio/http_plugin/beast_http_session.hpp (hunk 1 above)
Comment (spoonincode) [reply plugins/http_plugin/include/eosio/http_plugin/beast_http_session.hpp:257]: I usually always copy `shared_ptr`s, and it feels weird to move something in the initializer list that isn't (and can't be) moved in the ctor call,
https://github.com/AntelopeIO/leap/blob/5fcf5b49548e58e8bb01b50b6d7f2a9458602b0f/plugins/http_plugin/http_plugin.cpp#L230-L232

But the code generation is surprisingly better when adding the `move` here, so maybe it's worth doing

Comment (heifner) [reply plugins/http_plugin/include/eosio/http_plugin/beast_http_session.hpp:257]: Move of `shared_ptr` is nice as it avoid the `std::atomic` operation I believe.

**Review by greg7mdp — APPROVED** [review]

**Review by heifner — APPROVED** [review]

**Comment by ericpassmore** [comment]
Note:start
group: STABILITY
category: BUG
summary: Prevent error on shutdown by extending the life of `http_plugin_state` thereby ensuring no invalid memory accesses.
Note:end

**Pull request merged**